File: app/src/test/java/br/com/jonathanzanella/myexpenses/account/AccountRepositoryUnitTest.java
package br.com.jonathanzanella.myexpenses.account;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import br.com.jonathanzanella.myexpenses.database.RepositoryMock;
import br.com.jonathanzanella.myexpenses.validations.OperationResult;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Mockito.when;
import static org.mockito.MockitoAnnotations.initMocks;

public class AccountRepositoryUnitTest {
	@Mock
	private Account account;

	private AccountRepository accountRepository;

	@Before
	public void setUp() throws Exception {
		initMocks(this);
		accountRepository = new AccountRepository(new RepositoryMock<Account>());
	}

	@Test
	public void return_success_when_tried_to_save_valid_account() throws Exception {
		when(account.getName()).thenReturn("account");

		OperationResult result = accountRepository.save(account);

		assertTrue(result.isValid());
	}

	@Test
	public void return_with_error_when_tried_to_save_account_without_name() throws Exception {
		when(account.getName()).thenReturn(null);

		OperationResult result = accountRepository.save(account);

		assertFalse(result.isValid());
		assertTrue(result.getErrors().contains(ValidationError.NAME));
	}
}

File: app/src/test/java/br/com/jonathanzanella/myexpenses/unit/account/AccountRepositoryUnitTest.kt
package br.com.jonathanzanella.myexpenses.unit.account

import br.com.jonathanzanella.myexpenses.account.Account
import br.com.jonathanzanella.myexpenses.account.AccountDao
import br.com.jonathanzanella.myexpenses.account.AccountRepository
import br.com.jonathanzanella.myexpenses.validations.ValidationError
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import junit.framework.Assert.assertFalse
import junit.framework.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.Mock

class AccountRepositoryUnitTest {
    @Mock
    private lateinit var account: Account
    @Mock
    private lateinit var accountDao: AccountDao

    private lateinit var accountRepository: AccountRepository

    @Before
    @Throws(Exception::class)
    fun setUp() {
        account = mock()
        accountDao = mock()
        accountRepository = AccountRepository(accountDao)
    }

    @Test
    @Throws(Exception::class)
    fun return_success_when_tried_to_save_valid_account() {
        whenever(account.name).thenReturn("name")
        whenever(account.uuid).thenReturn("uuid")

        val result = accountRepository.save(account).blockingFirst()

        assertTrue(result.isValid)
    }

    @Test
    @Throws(Exception::class)
    fun return_with_error_when_tried_to_save_account_without_name() {
        whenever(account.name).thenReturn(null)
        whenever(account.uuid).thenReturn("uuid")

        val result = accountRepository.save(account).blockingFirst()

        assertFalse(result.isValid)
        assertTrue(result.errors.contains(ValidationError.NAME))
    }
}